cpp/common/src/codingstandards/cpp/types/Compatible.qll

```diff
@@ -353,45 +362,45 @@ module TypeEquivalence<TypeEquivalenceSig Config, interestedInEquality/2 interes
       or
       // We resolve typedefs, and one of these types is a typedef type: recurse.
       Config::resolveTypedefs() and
-      t1 instanceof TypedefType
-      or
-      t2 instanceof TypedefType
+      (
+        t1 instanceof TypedefType
+        or
+        t2 instanceof TypedefType
+      )
     )
   }
 
   /**
    * Check whether two types are equivalent, as defined by the `TypeEquivalenceSig` module.
    *
-   * This only holds if the specified predicate `interestedIn` holds for the types, and always
-   * holds if `t1` and `t2` are identical.
+   * This only holds if the specified predicate `interestedIn` holds for the types, or
+   * `interestedInNestedTypes` holds for the types, and holds if `t1` and `t2` are identical,
+   * regardless of how `TypeEquivalenceSig` is defined.
    */
   predicate equalTypes(Type t1, Type t2) {
-    compares(pragma[only_bind_into](t1), pragma[only_bind_into](t2)) and
+    interestedInNestedTypes(pragma[only_bind_into](t1), pragma[only_bind_into](t2)) and
     (
       t1 = t2
       or
       not t1 = t2 and
       if Config::overrideTypeComparison(t1, t2, _)
       then Config::overrideTypeComparison(t1, t2, true)
-      else
-        if t1 instanceof TypedefType and Config::resolveTypedefs()
-        then equalTypes(t1.(TypedefType).getBaseType(), t2)
-        else
-          if t2 instanceof TypedefType and Config::resolveTypedefs()
-          then equalTypes(t1, t2.(TypedefType).getBaseType())
-          else (
-            not t1 instanceof DerivedType and
-            not t2 instanceof DerivedType and
-            not t1 instanceof TypedefType and
-            not t2 instanceof TypedefType and
-            equalLeafRelation(t1, t2)
-            or
-            equalDerivedTypes(t1, t2)
-            or
-            equalTypedefTypes(t1, t2)
-            or
-            equalFunctionTypes(t1, t2)
-          )
+      else (
+        equalLeafRelation(t1, t2)
+        or
+        equalDerivedTypes(t1, t2)
+        or
+        equalTypedefTypes(t1, t2)
```

Comment:
`TypeEquivalenceSig::equalTypeDefTypes` has the following comment:
```
   * This predicate is only used if `resolveTypedefs()` is false. If so, then we assume two
   * typedefs are the same if they have the same name and their base types are equal.
```

So I think this case should only hold if `Config::resolveTypedefs()` is `false` - otherwise the resolution will be irrelevant.